### ISSUE TITLE
Standardize URI?

### ISSUE THREAD

**Issue by egberts** [canarytail/standard#4]
I read an article over at Hacker News (YCombinator.com) and then this current version of README.md; perhaps we also need a convergence upon a standardized mime:

* text/json 

And standardize URI like:

* canary.json

And URL like:
* http://example.invalid/canary.json
* http://example.invalid/.well-known/canary.json
* http://example.invalid/canary.html

I’m leaning toward something like:

* https://example.invalid/.well-known/canary.json

Or is it necessary to avoid a well known URI for this purpose?

**Comment by xAptive** [comment]
I had thought the same.  I think using .well-known is probably _right_ way to do it.  There is an [RFC](https://tools.ietf.org/html/rfc8615) covering it, as well as a (registration process)[https://www.iana.org/assignments/well-known-uris/well-known-uris.xhtml].  canary.json appears to be available.

**Comment by carrotcypher** [comment]
That does make sense. Building a PoC at the moment, will incorporate that into it as a test and update.

**Comment by egberts** [comment]
Ummm, just one more thing to consider, if you have plans in the future to do multiple files, you should tack on a ‘canary’ subdirectory after that `.well-known` directory part. 

But I highly suspect that won’t be the case.

**Comment by carrotcypher** [comment]
> Ummm, just one more thing to consider, if you have plans in the future to do multiple files, you should tack on a ‘canary’ subdirectory after that `.well-known` directory part.
> 
> But I highly suspect that won’t be the case.

When you say multiple files, are you referring to having multiple canaries from the same entity? I envision them having just one canary.txt (or whatever extension), but later on either: (1) having a network of storage (IPFS? TahoeLAFS? OrbitDB?) to keep track of the canaries; and/or (2) some kind of canary history inside the canary itself to cryptographically prove the last (few?) states.

At the moment I'm keeping things as simple as I can but this PoC being worked on will be a good place to jump in with ideas and I welcome to help.

**Comment by egberts** [comment]
K.I.S.S. is good.

Just trying to ensure no pitfalls with forward thinking.  Apparently, a web server by domain name should suffice ... for now.

**Comment by carrotcypher** [comment]
One problem I'm noticing from this is that if we keep the standard as it is now (with header/footer for easy cryptographic verification) then it's not valid JSON and basically just a text file that contains some JSON elements.

If we however move the cryprographic elements inside of the JSON array itself, I'm starting to notice that we would be complicating things exponentially (see: https://stackoverflow.com/questions/4670494/how-to-cryptographically-hash-a-json-object).

**Comment by egberts** [comment]
Agreed.  We can not insert comments into JSON (except as an object). 

Perhaps the simplified INI configuration format (or Python script full of assignments, #headduck)

**Comment by carrotcypher** [comment]
@egberts what do you think about just collapsing all the data fields together into a single string and signing that?

example:

The base information is

    {
    	"DOMAIN": "www.example.com",
    	"PUBKEY": "1GQH1cDFHLyq2KHHAqFTYMy9kWmMw9dsdR",
    	"NEWPUBKEY": "1DEKZAMjiT1jUhWAh8RH6zwmQpuvtW6X3y",
    	"PANICKEY": "1MKoEjCVyLcVrNZJEdxy4Utv4enNe5XNUn",
    	"NEWPANICKEY": "",
    	"VERSION": "0.1",
    	"RELEASE": "2019-03-06T22:23:09.963",
    	"EXPIRY": "2019-04-06T22:23:09.963",
    	"FRESHNESS": "0000 0000 0000 0000 000e d66a e55a 308b 77e5 ff78 bc94 a69b 8c27 da81 9562 537b",
    	"CODES": ["WAR", "GAG", "SUBP", "TRAP", "CEASE", "DURESS", "RAID", "SEIZE", "XCRED", "XOPERS", "SEPPU"]
    }

and we add a field called "SIGNED" which contains the signature of

    	"DOMAIN": "www.example.com","PUBKEY": "1GQH1cDFHLyq2KHHAqFTYMy9kWmMw9dsdR","NEWPUBKEY": "1DEKZAMjiT1jUhWAh8RH6zwmQpuvtW6X3y","PANICKEY": "1MKoEjCVyLcVrNZJEdxy4Utv4enNe5XNUn","NEWPANICKEY": "","VERSION": "0.1","RELEASE": "2019-03-06T22:23:09.963","EXPIRY": "2019-04-06T22:23:09.963","FRESHNESS": "0000 0000 0000 0000 000e d66a e55a 308b 77e5 ff78 bc94 a69b 8c27 da81 9562 537b","CODES": ["WAR", "GAG", "SUBP", "TRAP", "CEASE", "DURESS", "RAID", "SEIZE", "XCRED", "XOPERS", "SEPPU"]

which looks like

        "SIGNED":"1HZwkjkeaoZfTSaJxDw6aKkxp45agDiEzNG5ggR44agGR90xngVZvpqGKfgcQUPWyQmY93BfPUnfKkFpV+u6NC3Q/Q8i6ID2dY2oGYW2gsgz6sfLJrUlVz4A="

**Comment by carrotcypher** [comment]
Perhaps for clarity as a standard it should even be:


    {
        "CANARY": {
            "DOMAIN": "www.example.com",
            "PUBKEY": "1GQH1cDFHLyq2KHHAqFTYMy9kWmMw9dsdR",
    	    "NEWPUBKEY": "1DEKZAMjiT1jUhWAh8RH6zwmQpuvtW6X3y",
            "PANICKEY": "1MKoEjCVyLcVrNZJEdxy4Utv4enNe5XNUn",
            "NEWPANICKEY": "",
            "VERSION": "0.1",
            "RELEASE": "2019-03-06T22:23:09.963",
            "EXPIRY": "2019-04-06T22:23:09.963",
            "FRESHNESS": "0000 0000 0000 0000 000e d66a e55a 308b 77e5 ff78 bc94 a69b 8c27 da81 9562 537b",
            "CODES": ["WAR", "GAG", "SUBP", "TRAP", "CEASE", "DURESS", "RAID", "SEIZE", "XCRED", "XOPERS", "SEPPU"]
        },
        "SIGNED":"1HZwkjkeaoZfTSaJxDw6aKkxp45agDiEzNG5ggR44agGR90xngVZvpqGKfgcQUPWyQmY93BfPUnfKkFpV+u6NC3Q/Q8i6ID2dY2oGYW2gsgz6sfLJrUlVz4A="
    }

**Comment by egberts** [comment]
The only issue I could think of with self-enclosure of PGP keys inside a JSON is that it would require an API/toolkit to ensure proper and accurate text coverages for key signing within the ever-shifting JSON fields to obtain its concise and recreateable key value.  

Do we assign textual coverage (needed for key algorithmic computation by which starting braces of the nesting curly braces of JSON or do we cover just the JSON field name and value? What to do with reordering of JSON field names?

**Comment by egberts** [comment]
Dearmouring the canary file to reach the canary file is easy, consistent, and safe. 

Fishing out the “SiGNED” from a JSON then figuring out which JSON format to use ... not so much.

**Comment by carrotcypher** [comment]
Just to clarify, it wouldn't be PGP/GPG but just ECDSA (ed25519 curve). As for the order, since the signature would be of the fields as they are laid out by the author, and no one will be reordering the JSON when loading it into the CLI, it shouldn't be an issue to guarantee the order of the fields is always the same when verifying.

**Comment by egberts** [comment]
Embedding a SIGNED field name and its value would require some overt declaration of where to start reading and stop reading the cipher block to obtain/verify such SIGNED value.

**Comment by carrotcypher** [comment]
The way I was thinking, the CLI would process the JSON by:

1. grabbing the key to verify against from [CANARY][PUBKEY]
2. taking the entire contents of [CANARY] and flattening it into a single line (let's called it FLAT_CANARY)
3. confirming that the signature contained in [SIGNED] matches the integrity of FLAT_CANARY

I'll work on a PoC this month and post it to the client repo. If I'm missing something, I'll find out then! :)

**Comment by egberts** [comment]
Very Pythonese.  Probably JavaScriptist too.

**Comment by carrotcypher** [comment]
The standard has been updated to solve the signing problem. Check it out! https://github.com/canarytail/standard/blob/master/README.md#format

**Comment by egberts** [comment]
> Just to clarify, it wouldn't be PGP/GPG but just ECDSA (ed25519 curve). 

Uh?  By using an ECDSA, the keys would then become symmetric (unattributable and unchainable, yet, still verifiable; is this the design goal to somehow distribute the private key out-of-band un/securely that is needed for verification and validation process)?

GnuPGP/PGP is an asymmetric key management and has a verifiable, chainable, and attributable public key.

**Comment by carrotcypher** [comment]
> > Just to clarify, it wouldn't be PGP/GPG but just ECDSA (ed25519 curve). 
> 
> Uh?  By using an ECDSA, the keys would then become symmetric (unattributable and unchainable, yet, still verifiable; is this the design goal to somehow distribute the private key out-of-band un/securely that is needed for verification and validation process)?
> 
> GnuPGP/PGP is an asymmetric key management and has a verifiable, chainable, and attributable public key.

ECDSA is asymmetric. The usecase here is to allow for canary generators to generate a private key (for controlling and signing the canary), while sharing their public key for verification.

Adding GPG would of course add more features (like Web-of-trust) but also overhead and compatibility issues.

**Comment by snex** [comment]
I have an example up on my personal homepage, https://home.xens.org. The code checks a default location of "canary/canary.json" but it will also read a custom location from a META tag.

**Comment by egberts** [comment]
More on JSON and canary

https://news.ycombinator.com/item?id=42990948
